Implement pipeline rule timers that can be enabled/disabled at runtime

bernd opened the pull request:
## Description

Not having timers for pipeline rules makes it really hard to debug performance problems. This change adds a new interpreter listener that tracks rule evaluation (`when` statement) and execution (`then` statement) timers.

Since tracking timers for lots of rules can be impact processing performance, the timers are disabled by default. Users can toggle the rule timers on the rules management page.

Closes #5494

Hint: Review with "Hide whitespace changes" to make the diff smaller

##  Example Metric Names

```
org.graylog.plugins.pipelineprocessor.ast.Rule.5d5161cb2c9287782b2826d5.trace.evaluate.duration
org.graylog.plugins.pipelineprocessor.ast.Rule.5d5161cb2c9287782b2826d5.58fe004a8ae4922ab869ab53.0.trace.evaluate.duration

org.graylog.plugins.pipelineprocessor.ast.Rule.5d5161cb2c9287782b2826d5.trace.execute.duration
org.graylog.plugins.pipelineprocessor.ast.Rule.5d5161cb2c9287782b2826d5.58fe004a8ae4922ab869ab53.0.trace.execute.duration

```

## UI Configuration Screenshots

![image](https://user-images.githubusercontent.com/461/67483975-559e1a80-f667-11e9-8c60-0aa0de375755.png)

![image](https://user-images.githubusercontent.com/461/67483983-5e8eec00-f667-11e9-903a-6d468f6865a6.png)

![image](https://user-images.githubusercontent.com/461/67483997-677fbd80-f667-11e9-874e-3cca62b32807.png)

bernd commented:
@kroepke Is this approach similar to what you had in mind? In #5494 we talked about using sampling to avoid spending too much cycles on the timers. With this PR we maintain timers for all messages but the user can enable/disable the metrics collection at runtime. (disabled by default)

@edmundoa I would love to get your feedback on the UI changes. :smiley:

kroepke commented:
I think a global toggle is acceptable for now, given that it is only meant for debugging at the moment.
I'll add a review for some more for some specific questions/comments.

bernd commented on graylog2-web-interface/src/components/rules/RuleMetricsConfigContainer.jsx:
```diff
@@ -0,0 +1,60 @@
+  };
+
+  componentDidMount() {
+    RulesActions.loadMetricsConfig().then(this.showConfig).catch(this.showConfig);
```
Done.

bernd commented on graylog2-web-interface/src/components/rules/RuleMetricsConfig.jsx:
```diff
@@ -0,0 +1,115 @@
+                 checked={!nextConfig.metrics_enabled} />
+        </fieldset>
+        <p>
+          When enabled the system metrics will update two timers for every rule execution.
```
I thought about that. The problem is that the metrics route needs a node ID so we would have to decide which node to link to.

bernd commented on graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/processors/PipelineInterpreter.java:
```diff
@@ -93,6 +97,9 @@ public PipelineInterpreter(Journal journal,
         try (Timer.Context ignored = executionTime.time()) {
             final State latestState = stateUpdater.getLatestState();
+            if (latestState.enableRuleMetrics()) {
+                return process(messages, new RuleMetricsListener(metricRegistry), latestState);
```
Using a `LocalMetricRegistry` doesn't help in this case. We cannot reliably remove the timer metrics once the user disabled the metric collection because we don't have a synchronization point where all processing threads use the updated state.

That means even if we remove the timer metrics after disabling the collection, we don't know if one thread is still using the old state and re-creates the metrics again.

The metrics are cleaned up when a rule gets deleted or the server gets restarted.

bernd commented on graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/processors/listeners/RuleMetricsListener.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    public static String getMetricName(String ruleId, Type type) {
+        return name(Rule.class, ruleId, "trace", type.toString().toLowerCase(Locale.US), "duration");
```
Will do. :+1:

bernd commented:
@kroepke @edmundoa Thanks for the comments!

I updated the PR accordingly:

- Create timer metrics for pipeline and stage IDs as well
- Updated UI as requested

bernd commented on graylog2-web-interface/src/components/rules/RuleMetricsConfig.jsx:
```diff
@@ -0,0 +1,118 @@
+  };
+
+  constructor(props) {
+    super(props)
```
Thank you! I will fix that.
